core/src/main/java/org/apache/calcite/avatica/Meta.java

```diff
@@ -685,7 +686,14 @@ public static CursorFactory deduce(List<ColumnMetaData> columns,
       if (List.class.isAssignableFrom(resultClazz)) {
         return LIST;
       }
-      return record(resultClazz);
+      // columns for MetaImpl internal classes do not match field names, "record" would fail
+      // can be removed once CALCITE-2489 is merged in Calcite
+      if (resultClazz.getName().startsWith("org.apache.calcite.avatica.MetaImpl")
+          && !resultClazz.getName().endsWith("MetaTable")) {
+        return ARRAY;
+      }
```

Comment:
Also, probably not our concern, but just for the record: this piece of code might not work in a downstream project if shading is involved

Reply:
With or without the if statement the `deduce` method changes behavior and it might break downstream projects (as it breaks Calcite for instance). I think it is worth adding a backward compatibility note for the method and the change in general in `history.md`.

Reply:
I was not sure if such a "breaking change" was acceptable, in that case I agree we can let Calcite adapt to the change. This allows us to avoid the shading issue @rubenada mentions, thanks for the suggestion btw.

For `history.md` a note along these lines could work?

> `Meta#deduce(List<ColumnMetaData> columns, Class resultClazz)` now derives the order of the fields from the list of provided column metadata `columns` when generating a record from the given Java class `resultClazz`, instead of relying on the field order provided by `Object#getFields()`, which is a JVM-dependent feature.

`history.md` seems to be updated during the release itself by combining the list of commits and associated comments. I don't know what's the usual way for adding such a note, maybe it can be part of the commit message itself?

Reply:
Usually when there is a breaking change people update `history.md` directly when they commit the change. During the release the release manager will make a pass on `history.md` to add commits etc. and reorganise/reword the content as they see fit.

The message looks good although it doesn't explain for instance that previous working code might now throw something like the following:  `java.lang.RuntimeException: java.lang.NoSuchFieldException: TABLE_CAT` Why this happens and how should they fix it?

Reply:
Thanks @zabetak, I have added a new commit with an update to `history.md` with the explanation of the change. I haven't included the changes from your branch, although they all LGTM.

Reply:
I have also added javadoc to `deduce` to cover explicitly the expectations for columns and resultclazz, and that an exception is thrown on validation failure.

Reply:
Adding javadoc is a very good idea, thanks @asolimando !